terraform/provider/resource_teleport_role.go

```diff
@@ -1,61 +1,145 @@
 package provider
 
 import (
-	"github.com/gravitational/teleport-plugins/terraform/tfschema"
+	"context"
 
+	"github.com/gravitational/teleport-plugins/terraform/tfschema"
+	"github.com/gravitational/trace"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	roleNamePath = "metadata.0.name"
 )
 
+// teleport_role resource definition
 func resourceTeleportRole() *schema.Resource {
 	return &schema.Resource{
-		Create: resourceTeleportRoleUpsert,
-		Read:   resourceTeleportRoleRead,
-		Update: resourceTeleportRoleUpsert,
-		Delete: resourceTeleportRoleDelete,
-		Exists: resourceTeleportRoleExists,
+		CreateContext: resourceRoleCreate,
+		ReadContext:   resourceRoleRead,
+		UpdateContext: resourceRoleUpdate,
+		DeleteContext: resourceRoleDelete,
 
 		Schema: tfschema.SchemaRoleV3(),
 	}
 }
 
-func resourceTeleportRoleUpsert(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*client.Client)
-	name := d.Get("name").(string)
+// resourceRoleCreate creates Teleport role
```

Comment:
Nit: Could you please try to make the Godocs a bit more helpful to the reader? I.e. creates a Teleport role from what, where, etc.

Reply:
Okay. I am not sure I could add a lot of details here: it's standard Terraform CRUD, possibly it would be better to provide a link to Terraform docs somewhere in this file?